File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Easy/74. Search a 2D Matrix/C++ solution/74. Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int ROWS = matrix.size(), COLS = matrix[0].size();
        int L = 0, R = ROWS * COLS - 1;
        while (L <= R)
        {
            int mid = L + (R - L) / 2; // To avoid any over flows
            int row = mid / COLS, col = mid % COLS;
            
            if (matrix[row][col] == target)
                return true;
            else if (matrix[row][col] > target)
                R = mid - 1;
            else
                L = mid + 1;
        }
        return false; // If not found
    }
};

File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Medium/74. Search a 2D Matrix/Python solution/74. Search a 2D Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROWS, COLS = len(matrix), len(matrix[0])
        L, R = 0, ROWS * COLS - 1
        
        while L <= R:
            mid = L + (R - L) // 2 # this way we can avoide any overflows
            row = mid // COLS 
            col = mid % COLS
            if matrix[row][col] == target:
                return True 
            elif matrix[row][col] > target:
                R = mid - 1
            else:
                L = mid + 1
        
        return False # If not found

# NOTE
'''
The intuition is: First visualise the matrix as if it was a list of 1D numbers and
 then run BS on it, after that determine the location of that number in the
 2D matrix using mid // COLS to get the row of the number in the 2D and
 mid % COLS to get the column number of that number. The rest is the same as that of BS.
'''